### PR TITLE
Add minSize to circle.ts and rect.ts to prevent drawing unintentionally tiny shapes

### PR DESCRIPTION
**Pull request by hiankun**
I found it might be more reasonable to limit the circle and the rectangle to their minimal size by the width of stroke width.
Without the limitation, I have been occasionally added some "very tiny" circles or rectangles on the canvas and didn't aware it.

I have intermediate Python experience but nothing about web apps nor typescript (very beginner understanding about JS).
Therefore I tried to modify the code by guessing the rules from the original code base, and there are probably some bad programming styles or practices.

Also this is my first PR. I've tried to get rid of the unwanted files (i.e., the folder named `test_app/`) but got no luck.
In short, only `circle.ts` and `rect.ts` are relevant in this PR.
If there are anything I can study from please also let me know.

### CODE REVIEW

File: streamlit_drawable_canvas/frontend/src/lib/rect.ts [hunk 1]
```diff
@@ -71,8 +73,13 @@ class RectTool extends FabricTool {
...
     }
-    this.currentRect.set({ width: Math.abs(this.currentStartX - pointer.x) })
-    this.currentRect.set({ height: Math.abs(this.currentStartY - pointer.y) })
+    let minSize = this.minSize
+    let _width = Math.abs(this.currentStartX - pointer.x);
+    let _height = Math.abs(this.currentStartY - pointer.y);
+    if (_width <= minSize) { _width = minSize }
+    if (_height <= minSize) { _height = minSize }
+    this.currentRect.set({ width: _width })
+    this.currentRect.set({ height: _height })
```
Comment (andfanilo) [inline streamlit_drawable_canvas/frontend/src/lib/rect.ts:82]: Same as above

Could you do it more declaratively with `this.currentRect.set({ width: Math.max(_width, this.minSize) })` instead?

* Also I think there is no need for `let minSize = this.minSize`, the code is simple enough to know that we can't inadvertedly mutate the value :)

File: streamlit_drawable_canvas/frontend/src/lib/circle.ts [hunk 2]
```diff
@@ -22,6 +23,7 @@ class CircleTool extends FabricTool {
     this.strokeWidth = strokeWidth
     this.strokeColor = strokeColor
     this.fillColor = fillColor
+    this.minSize = strokeWidth
```
Comment (andfanilo) [inline streamlit_drawable_canvas/frontend/src/lib/circle.ts:26]: I don't think this is going to be used outside the class right? We can make it private if that's the case, by prepending with an underscore, so `this._minsize` and `_minsize: number` in the declaration

The other arguments are shared between multiple classes, which is why they are public.

Also if we make this private, we can call it even more explicitly `_minRadius` instead :wink:

File: streamlit_drawable_canvas/frontend/src/lib/rect.ts [hunk 3]
```diff
@@ -22,6 +23,7 @@ class RectTool extends FabricTool {
     this.strokeWidth = strokeWidth
     this.strokeColor = strokeColor
     this.fillColor = fillColor
+    this.minSize = strokeWidth
```
Comment (andfanilo) [inline streamlit_drawable_canvas/frontend/src/lib/rect.ts:26]: Same as above

I don't think this is going to be used outside the class right? We can make it private if that's the case, by prepending with an underscore, so `this._minsize` and `_minsize: number` in the declaration

If we make this private, let's rename it more explicitly to `minLength`

**Review by andfanilo — COMMENTED** [review]
Hey @hiankun, that sounds like a good idea thanks! The PR is great, I think you got the correct code modifications, I just have some nitpicks:

* As you mention, I think I can remove the `test_app` and `package-lock.json` modifications from the PR before merging, but if you're able to revert the changes before I'd be grateful :).
* You can use `Math.max(value, minValue)` directly instead of the `if value < minValue then value = minValue` routine, so we are not too imperative on this part
* You can make those `minSize` args private since they are not shared between all the classes. And so by making them private we can rename them more explictly to `minRadius` for example! 
* Did you test what happens when you transform a already created circle? Can you transform the circle to a radius lower than the `minSize`? I don't think it's a problem though, just pure curiosity.

Otherwise great job overall, hope we can get this merged soon

Fanilo

**Comment by hiankun** [comment]
Wow, thank you so much for such a reply with so many information. I believe that your suggestions are clear but I just haven't had enough Git and Typescript knowledge to get them quickly. I will study them and please bear with me and my future questions.

File: streamlit_drawable_canvas/frontend/src/lib/circle.ts [hunk 4]
```diff
@@ -9,6 +9,7 @@ class CircleTool extends FabricTool {
   currentCircle: fabric.Circle = new fabric.Circle()
   currentStartX: number = 0
   currentStartY: number = 0
+  minSize: number = this.strokeWidth
```
[the remark this answers is not in the document]
Comment (hiankun) [reply streamlit_drawable_canvas/frontend/src/lib/circle.ts:12]: In Ln 7, there's `strokeWidth: number = 10`. I thought it was similar to the initial settings in Python class, and can be accessed by using `this.strokeWidth`. By doing so, I could express the idea that "the minSize is meant to be the same as the strokeWidth". If it's not the case, then indeed I should just set it to `0` or something.

**Comment by hiankun** [comment]
I think I have finished the code with slightly more consistent logic and style (thanks for the suggestions again).

Unfortunately I also messed up my forked repo.

Should I close this and open a new one so that I can give a cleaner PR? (I probably would re-fork a new repo to avoid all the mess...)

**Comment by hiankun** [comment]
To the question of transforming a created circle: Yes, the transform will not be limited by the minimal size.

(I also noticed that the transform don't preserve the stroke width. If possible, it would be nice to have an option to keep the stroke width while transforming the shapes. Of course it would be another topic.)

**Comment by andfanilo** [comment]
> Unfortunately I also messed up my forked repo.
> 
> Should I close this and open a new one so that I can give a cleaner PR? (I probably would re-fork a new repo to avoid all the mess...)

If you feel courageous, you could create a new branch on your forked repo from the base branch (before your own commits) and merge in the files from your other branch, then recreate/edit the branch to merge in the PR..

If you feel less courageous, then yeah reforking, overwriting with the correct changes and closing -> recreating a PR is totally ok!

**Comment by hiankun** [comment]
Hi, thank you for the workflow guidance, which has been what I need. I think I have fixed the PR (as it now shows only the two changed files). Please take a look and let me know whether I did it right or not.

**Comment by andfanilo** [comment]
Hello @hiankun 
Finally had a look at your changes locally and it looks good!

I'm ready to merge, but for this we need 2 last things:
* Could you format your code with Prettier? I usually have VSCode do it on Save as you write, but if you're using another IDE then doing `npx prettier --write src\lib\rect.ts` and `npx prettier --write src\lib\circle.ts` at `streamlit_drawable_canvas\frontend` level should do
* Your PR is still in draft mode so I can't merge it :upside_down_face: 

Have a nice day,
Fanilo :balloon:

**Comment by hiankun** [comment]
Hi @andfanilo ,

Thanks again for all the suggestions and instructions from which I've learned a lot. :-D
